### ISSUE TITLE
Kafka consumer subscribers -- member assignments being lost (nodejs)

### ISSUE THREAD

**Issue by BryanDollery** [tulios/kafkajs#1328]
### Synopsis
We have a subscriber that responds to events on 5 topics with a single consumer-group, each with a single partition. When firing events into these topics, we see our code handling the first two events, then nothing. When we check, on startup, the consumer group has 5 member assignments, but after handling the 2nd event *all* of the member assignments disappear. 

### Code
Here is the subscription code:
```js
  connectoToConsumer: async () => {
    const groupId = `${Group}-adaptor`;
    log.debug({ msg: `Connecting to event consumer...`, groupId });
    _app.connection.consumer = _app.kafka.consumer({ groupId });
    await _app.connection.consumer.connect();
  },

  subscribe: async topics => {
    log.debug("Subscribing to many topics...");
    for (const topic of topics.split(",")) {
      await _app.subscribeConsumerToTopic(topic);
    }
  },

  subscribeConsumerToTopic: async topic => {
    log.debug({ msg: "Subscribing to topic", topic });
    try {
      await _app.connection.consumer.subscribe({ topic: topic, fromBeginning: false });
    } catch (e) {
      log.error({ msg: `FATAL ERROR: Event adaptor consumer connection failed`, type: "error", subtype: "fatal", source: "registerConsumer", cause: e });
      process.kill(process.pid, "SIGINT");
    }
  },

  runConsumer: async () => {
    await _app.connection.consumer.run({
      autoCommitInterval: 100,
      autoCommitThreshold: 100,
      eachMessage: async ({ ignore_topic, ignore_partition, message: event }) => handler(event)
    });
  },
```

To invoke this we use:
```js
await _app.connectoToConsumer();
await _app.subscribe(topics);
await _app.runConsumer();
```

### Output
When we run the consumer we see a log message from kafka:

```json
{
  "level": "INFO",
  "timestamp": "2022-04-21T11:23:30.186Z",
  "logger": "kafkajs",
  "message": "[ConsumerGroup] Consumer has joined the group",
  "groupId": "service-invoice.projection.1-adaptor",
  "memberId": "ctx-service-invoice.projection.1-adaptor-f3ff61a9-9d21-4aeb-9560-ba415188cacd-3914c6d0-6674-4342-8a45-8114c3cd6b62",
  "leaderId": "ctx-service-invoice.projection.1-adaptor-f3ff61a9-9d21-4aeb-9560-ba415188cacd-3914c6d0-6674-4342-8a45-8114c3cd6b62",
  "isLeader": true,
  "memberAssignment": {
    "service-invoice.created.1": [
      0
    ],
    "service-invoice.approved.1": [
      0
    ],
    "service-invoice.service-orders-created.1": [
      0
    ],
    "service-invoice.service-orders-invoiced.1": [
      0
    ],
    "service-invoice.customer-changed.1": [
      0
    ]
  },
  "groupProtocol": "RoundRobinAssigner",
  "duration": 23555
}
```

Showing that we have successfully subscribed to 5 topics. We expect to see those 5 events in the order listed above. However, after the `approve` event, no other events are handled, and when we look at kafka all the members have gone. There is no exception or other log that would suggest an error in our code.

Incidentally, our handlers run very fast. They make a local HTTP call to another container in the same pod that updates a single field in a single document in MongoDB. Typically the handlers run and respond in < 500ms.

### Suspicions
As JavaTechnical suggests in the answer below, we believe that this may be related to the heartbeat mechanism. The kafka logs show this:

```log
2022-04-21 13:17:56,941 INFO [GroupCoordinator 0]: Member ctx-service-invoice.projection.1-adaptor-4b32ab43-17f2-4b98-84ed-b21ee36b771f-d6040fbd-2fb8-4025-9fdd-5fb8fb6b41a1 in group service-invoice.projection.1-adaptor has failed, removing it from the group (kafka.coordinator.group.GroupCoordinator) [executor-Heartbeat]
```

We're using strimzi to run kafka in k8s, and we see in the logs:

```log
broker.heartbeat.interval.ms = 2000
broker.session.timeout.ms = 9000
group.max.session.timeout.ms = 1800000
group.min.session.timeout.ms = 6000
```

We think that the consumption of the first two messages is probably just luck -- we're getting them before the session-timeout is triggered by the lack of heartbeats. 

### What have we tried?

* We tried setting the heartbeat interval in our kafkajs code:

```js
.connection.consumer = _app.kafka.consumer({ groupId, , heartbeatInterval: 1000 });
```

But that isn't helping. 

* Kafkajs documentation says that the `eachMessage` structure in the consumer runner should automate the heartbeats, but given the nature of the error we've also tried explicitly passing the heartbeat function to the handler and manually invoking it, also to no avail. 

* We've also tried extending the timeouts and heartbeat intervals using strimzi, but our attempts to do so failed (they crashed strimzi) so we've backed out of that.

### Question
Given all of this, we suspect that the kafkajs heartbeats are not making it through to kafka in k8s under strimzi. Our question is, how do we ensure that this works?

### Versions
We're using the latest strimzi, kafka, and kafkajs in node 17+.

### Reference
We have posted this question to Stack Overflow too.

**Comment by Nevon** [comment]
If you run KafkaJS with the [log level set to debug](https://kafka.js.org/docs/configuration#logging) you can see each request the client makes towards the broker, including heartbeats, which can help you confirm or rule out your suspicion that heartbeats are not happening frequently enough.

The heartbeating in `eachMessage` is pretty straightforward. It just loops over each message in the batch and heartbeats in between each one as long as at least `heartbeatInterval` milliseconds has elapsed since the last heartbeat.

https://github.com/tulios/kafkajs/blob/4f0c52de2c928dff0e2fb3afb550545de4e412c9/src/consumer/runner.js#L145-L178

The default `heartbeatInterval` is 3000ms, which is longer than what strimzi expects according to your `broker.heartbeat.interval.ms`, but if you've lowered it to 1000 then it shouldn't be an issue. Since we don't see the definition of `_app` here, are you sure that `_app.kafka.consumer` is forwarding all arguments to `kafka.consumer`?

**Comment by peteriman** [comment]
Might need to check the `session.timeout.ms`, as it is the property for broker to receive at least 1 heartbeat from client.

**Comment by BryanDollery** [comment]
@Nevon Sorry it has taken me so long to respond. I traced the code to the heartbeat too. I think it should be sent when there are no messages too -- my service is sitting idle most of the time and I think that's when the heartbeat needs to be sent. Unless I'm wrong about the function of a heartbeat?

**Comment by Nevon** [comment]
You are correct that heartbeats should be sent even when there are no messages, but this is done internally by KafkaJS. Somewhat simplified, it essentially has a loop where it sends fetch requests to the brokers to get new messages, then it passes those messages to the userland handlers for processing one by one, and in between it heartbeats (respecting the user's configuration in terms of frequency). If there are no messages to process, it sends a heartbeat and then restarts the loop.
